store/src/state.rs

```diff
@@ -39,9 +39,85 @@ pub struct TransactionInputs {
     pub nullifiers: Vec<NullifierInfo>,
 }
 
+/// Nullifier SMT.
+#[derive(Debug, Clone)]
+struct NullifierTree(Smt);
+
+impl NullifierTree {
+    /// Construct new nullifier tree from list of items.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (Nullifier, BlockNumber)>
+    ) -> Result<Self, NullifierTreeError> {
+        let leaves = entries.into_iter().map(|(nullifier, block_num)| {
+            (nullifier.inner(), Self::block_num_to_nullifier_value(block_num))
+        });
+
+        let inner = Smt::with_entries(leaves)?;
+
+        Ok(Self(inner))
+    }
+
+    /// Get SMT root.
+    pub fn root(&self) -> RpoDigest {
+        self.0.root()
+    }
+
+    /// Returns an opening of the leaf associated with the given nullifier.
+    pub fn open(
+        &self,
+        nullifier: &Nullifier,
+    ) -> SmtProof {
+        self.0.open(&nullifier.inner())
+    }
+
+    /// Inserts block number in which nullifier was consumed.
+    pub fn insert(
+        &mut self,
+        nullifier: &Nullifier,
+        block_num: BlockNumber,
+    ) -> Result<(), NullifierTreeError> {
+        let value = self.0.insert(nullifier.inner(), Self::block_num_to_nullifier_value(block_num));
+        if value == EMPTY_WORD {
+            return Ok(());
+        }
+
+        Err(NullifierTreeError::NullifierAlreadyExists {
+            nullifier: *nullifier,
+            block_num: Self::nullifier_value_to_block_num(value),
+        })
+    }
+
+    /// Returns block number stored for the given nullifier or `None` if the nullifier wasn't
+    /// consumed.
+    pub fn get_block_num(
+        &self,
+        nullifier: &Nullifier,
+    ) -> Option<BlockNumber> {
+        let value = self.0.get_value(&nullifier.inner());
+        if value == EMPTY_WORD {
+            return None;
+        }
+
+        Some(Self::nullifier_value_to_block_num(value))
+    }
+
+    /// Returns the nullifier's leaf value in the SMT by its block number.
+    fn block_num_to_nullifier_value(block: BlockNumber) -> Word {
+        [Felt::from(block), Felt::ZERO, Felt::ZERO, Felt::ZERO]
+    }
+
+    /// Given the leaf value of the nullifier SMT, returns the nullifier's block number.
+    ///
+    /// There are no nullifiers in the genesis block. The value zero is instead used to signal
+    /// absence of a value.
+    fn nullifier_value_to_block_num(value: Word) -> BlockNumber {
+        value[0].as_int().try_into().expect("invalid block number found in store")
+    }
```

Comment:
nit: I don't think "nullifier_value" is the right word here (block number is not the value of a nullifier). I'm thinking better names for these functions would be `block_num_to_leaf_value` and `leaf_value_to_block_num`.